### ISSUE TITLE
[2.0] Consider abandoning the polyfill

### ISSUE THREAD

**Issue by rtheunissen** [php-ds/ext-ds#92]
There are a few reasons why I believe we should consider abandoning the polyfill:
- It's a performance risk. Someone might not realise that they don't have the extension enabled or share code with someone that doesn't have the extension installed, which results in a very inefficient implementation of the structures.
- It's blocking some changes in the extension, like operator overloading on sets (for diff, xor, union and complement) and the ability to allow `==` to behave like it does for arrays.
- There isn't a good reason for anyone to use the polyfill on purpose. If they don't have access to their environment to install the extension, they shouldn't be using the library at all.
- It's extra work to maintain the polyfill, extension and tests in separate repositories. If we drop the polyfill we can move the tests into the extension repository and tag them together.

**Comment by rtheunissen** [comment]
One thing to note is that the polyfill also acts as an autocomplete source. If we decide to abandon the polyfill in 2.0, we would need to keep the stubs anyway. Something we could do is maintain a composer package that contains the tests and the stubs, and *requires* the extension as `ext-ds`, and have the extension repository require that package. That way we're still separating the extension from the tests and the stubs, and we're enforcing that the extension is installed if a project has the composer package as a dependency.

**Comment by kelunik** [comment]
Without a polyfill we (amphp) will never use `ds`. Things must just work for all basic things.

**Comment by rtheunissen** [comment]
@kelunik but you're taking two steps back by using the polyfill at all. Anyone without knowledge of the extension using amphp will suffer serious performance and memory issues.

**Comment by kelunik** [comment]
Right, we currently don't use it, but if we would, we'd want it to be runnable by default without any additional extensions. It's currently not worth for the few cases we have.

**Comment by stepzerosolutions** [comment]
With first installation attempt fail, I've installed the extension without any issue on a new php build. I'm new to php-ds and still doing testing, I don't see any reason not to abandoned it if nobody is using it.

**Comment by rtheunissen** [comment]
@TheCelavi @jkuchar @dantudor @krakjoe @nikic in light of this issue and #8 (Pure PHP implementation), I would like to ask if anyone has any strong opinions here. 

In my opinion:

*Reasons to abandon*:
  - No chance that the library decreases performance (either extension or not at all).
  - No need to maintain the second implementation - can focus on the extension.
  - There isn't a practical reason why anyone would be using the polyfill, eg. if you're in an environment where you don't have access to install the extension, you shouldn't be using the library at all.

*Reason to maintain*:
  - Improves adoption, can play around without installed the extension.
  - Allows code to run in an environment where the extension isn't installed.
  - Doubles as autocomplete stubs.
  - Most of it is already implemented.
  - We'll have to maintain it for 1.0 anyway.

**Comment by jkuchar** [comment]
It was big thing when I was considering if we are going to use phpds interface. For example when there will be new version of php and this extension will not be available (e.g. not maintained anymore) I will still be able to use the app thanks to the polyfill or even fix someting there. I will not be able to fix C source. Polyfill makes me feel much more safe that my app will not be locked on old php version where phpds used to be supported.

I use polyfill all the time because I do not want to mess with the compilation process on Windows. I do not care that it is slow. It works.

It makes sense to me to NOT have polyfill if phpds makes it into php bundled extensions. (e.g. just enable it in php.ini and go; or just install package and go on linux)

Honza Kuchař

23. 7. 2017 v 3:13, Rudi Theunissen <notifications@github.com>:

**Comment by rtheunissen** [comment]
@jkuchar I think what dawned on me is that you're better of either not upgrading PHP or not using the extension than you are using the polyfill. 

> I do not want to mess with the compilation process on Windows

You shouldn't need to at all, there are `.dll`'s in the releases.

**Comment by jkuchar** [comment]
I didn't know it made it into PHP 7.1 default bundle. Now it feels like something stable.

The mission of polyfill - to overcome the fear of getting dependent on new extension whose development can be abandoned - is completed.

25. 7. 2017 v 2:53, Rudi Theunissen <notifications@github.com>:

**Comment by rtheunissen** [comment]
> I didn't know it made it into PHP 7.1 default bundle

Not that I'm aware of, but it's compiled and available on PECL.

> The mission of polyfill - to overcome the fear of getting dependent on new extension whose development can be abandoned - is completed.

The mission of the polyfill is:
- Prevent code from not working *at all* in environments where the extension couldn't be installed or hasn't been installed yet
- Provide IDE autocomplete by having the method signatures in your /vendor
- Provide an easy way for people to "play with" the API without the need to install the extension.

But will all this said and done, I'm starting to feel like maybe failing hard when the extension is not available is a good thing. Because you'd be well worse off then than if you were using arrays.

I'm not convinced either way yet, so this is still just a consideration. 🤔

**Comment by NoodlesNZ** [comment]
I agree with abandoning the polyfill. If the performance is not there then it's not worth having (it defeats the purpose of using ds). Maintaining the polyfill takes effort away from the extension.

It shouldn't be that hard to maintain stubs for autocomplete.

**Comment by rtheunissen** [comment]
That's what I'm leaning towards at the moment. I fail to see what value it has going forward. Either use the extension or don't use the library at all.

**Comment by nikic** [comment]
How bad is the performance of the polyfill?

**Comment by rtheunissen** [comment]
It's not horrendous. The only operation that makes me nervous is put and get on Map being O(n). 

Haven't done any benchmarks.

**Comment by NoodlesNZ** [comment]
The polyfill test suite runs on my mac:
```
Time: 24.97 seconds, Memory: 20.00MB

OK (2790 tests, 7059 assertions)
```
extension:
```
Time: 1.14 seconds, Memory: 12.00MB

OK (2790 tests, 7068 assertions)
```

**Comment by nikic** [comment]
> The only operation that makes me nervous is put and get on Map being O(n).

OMG. That's really bad.

If that's the state of the polyfill implementation, I would definitely abandon it. On the other hand, with that issue resolved performance might be acceptable.

**Comment by rtheunissen** [comment]
I don't know if it's a solvable problem, unless we build a hashtable using an array, but the memory usage would be huge if we do that. Strong types on the keys and objects as keys makes it difficult to use an array as the internal store because we can't index. 

On top of that we still have to maintain insertion order.

**Comment by nikic** [comment]
@rtheunissen The current definition of the Hashable interface is what makes this hard. If Hashable::hash() would be required to uniquely identify an object (of the given class) it should be easy to implement something relatively efficient.

**Comment by rtheunissen** [comment]
@nikic that would make it easy yes, but that's not realistic here. We can't use `hash` for equality also, it's only to improve distribution of the keys.

**Comment by nikic** [comment]
@rtheunissen Given how people would implement this function in userland, I don't think it's unrealistic. As we don't expose hashing primitives (for strings in particular), the most obvious ways of implementing hash() are also going to be collision-free.

**Comment by rtheunissen** [comment]
For example, a database entity's primary key? I can see how the common implementation would double as an identity function. But I don't like the idea of changing the definition simply for the sake of accomodating the polyfill.

There's also the case of using an array as a key.. where the hash is now the length of the array, and get/put is O(k) because we have to compare the elements of the array. 

Intuitively I can't get my head around how we can do this efficiently in userland without a proper native array type that we can use to build a hashtable.

**Comment by nikic** [comment]
I wouldn't worry about performance of array keys, that's a very unusual use-case.

You can't do this *really* efficiently in userland, but implementing a userland HT should also not be much less memory efficient than what you have now (due to the general object overhead, adding two more fields on top of Pair is not going to double the memory usage).

But then again, implement anything here is going to take time and I can see how that time might be better spent on something else...

**Comment by rtheunissen** [comment]
Yeah I just don't see the point of going out of our way here.

**Comment by kelunik** [comment]
> Either use the extension or don't use the library at all.

That only works if the user of the library is also controlling the system where the app is to be deployed.

**Comment by rtheunissen** [comment]
@kelunik which is why I'm saying if you don't control the system, don't use the library. In saying that, if it's a dependency of another dependency, you're in trouble. 

But how is that any different to extensions like mongo or pthreads? 

My concern is that if you don't have control of the system now, you probably won't ever have, and you're then stuck with slow code under the hood. Could argue that it's then insignificant because you won't be processing a lot if you're not controlling your system anyway.

**Comment by kelunik** [comment]
> My concern is that if you don't have control of the system now, you probably won't ever have, and you're then stuck with slow code under the hood. Could argue that it's then insignificant because you won't be processing a lot if you're not controlling your system anyway.

Right, on those systems I / we don't care about performance, but pure portability and runability.

> But how is that any different to extensions like mongo or pthreads?

We abstract those where possible, e.g. using multiple processes instead of threads where they're not available in https://github.com/amphp/parallel. If you require mongo to run, you can also require the extension, because the user has to be in control then anyway to run mongo. But that's nothing desirable for generic data structures.

It definitely depends on the use case, but for some cases it's just easier using arrays or Spl then, even if they're worse implementation wise.

**Comment by rtheunissen** [comment]
At the end of the day, the only cost is development time, which I'm happy to provide. Let's just keep the polyfill until it causes actual, real world problems.

We're suggesting the extension when it's installed with Composer, and we can go as far as creating a warning when the package is loaded (which means the extension is missing).

**Comment by rtheunissen** [comment]
See #67 for a good case against the polyfill. In light of these discussions, it's my opinion that we should abandon the polyfill entirely. ✅

**Comment by morrisonlevi** [comment]
I've seen an increase in attempts to add generics to PHP core. I am working on parameterized traits myself. Have you already gone through to see how they would affect this library? Maybe porting the polyfill to Hack?

I'm bringing it up here because the polyfill seems a much nicer test-bed than C extensions are.

**Comment by rtheunissen** [comment]
As soon as they're in core, we'll evaluate how they could apply to this library. I don't believe that porting the polyfill to Hack would have much practical application. While it is a nicer test-bed than a C extension, it holds the extension back and has very little use outside of testing and playing around.

My current plan for 2.0 (which I'll get around to soon) is to make the polyfill stubs only, and move them and the tests into the same repo as the extension.

**Comment by morrisonlevi** [comment]
I don't understand the equality issue. We have a `Hashable` interface with an `equals` method (side note: we probably should have an `Equatable` interface with `equals` on it and have `Hasbable` extend it). Users should be using that API, not `==`. However, I do understand the value of keeping the polyfill around for experimenting with new features and API, which comes from having written and collaborated on many collection style libraries.

I don't think you understood my comment about Hack at all. I'm suggesting that inasmuch as it is possible *now* we make the library compatible with generics. This makes the BC breaks smaller *when* it comes. At this point I'm confident it's a matter of when and not if. You may not share that confidence, which is fine. I was just asking if you'd done any work in the area already and thought you might because Hack exists.

**Comment by mcneely** [comment]
I'm just curious as to why the polyfill doesn't use class_exist to skip installation declaration if the class is already declared?

**Comment by rtheunissen** [comment]
@mcneely what effect would that have / what problem would that solve?

**Comment by mcneely** [comment]
@rtheunissen Does is not raise a warning or error if the extension is already installed?

**Comment by rtheunissen** [comment]
If the extension is loaded, it won't try to autoload the polyfill at all. Not sure where the warning would come in. If you have the extension installed and you use `php -dextension` then it might warn that it's already installed.

**Comment by mcneely** [comment]
For some reason I had it in my head it would fail. I'm currently working on a personal project that incorporates the extension/polyfill and abandoning the polyfill would require me to yank the code altogether and not incorporate the extension at all
.

**Comment by rtheunissen** [comment]
@mcneely why would it require you to yank the code? Can you not specify the extension as a dependency, as you would PDO or json? The polyfill will still be maintained for 1.x but won't be supported for the next major version onwards.

**Comment by mcneely** [comment]
@rtheunissen Not everyone has the ability to install the extension. I'm trying to make a library that most can use, so unless 2.0 is getting compiled in I really don't see an option. I do see a lot of places where the polyfill could be optimized, however. For example, The map class could either possibly refactored to use the splObjectStorage as a base or at least use the key in the $pairs array so you don't have to loop through each item to find the key (if it's an object, doing something like a sha-1 on the serialized object as the key)

**Comment by rtheunissen** [comment]
There are some features in 2.0 that might not be able to be implemented in user PHP, or at least at some point that might be the case. SplObjectStorage isn't really an option because it accepts *only* objects, and we have to store pairs because arrays convert everything to string/int. I don't see a way to create a map that doesn't have linear time lookups.

Anyone who doesn't have the ability to install the extension should change their environment such that they can, otherwise they can't use your library. This would be people on shared hosts right? 

I doubt ds will ever be compiled in, or merged with core PHP. There is a chance but it's not in discussion at this stage. Someone using php-ds without the extension, or someone writing a library where that's a possibility, is doing themselves and all users a disservice. Performance will be worse than if you just used arrays.

**Comment by morrisonlevi** [comment]
> I don't see a way to create a map that doesn't have linear time lookups.

Hash + linear search for collisions only. Something like this [dict implementation](https://github.com/morrisonlevi/Ardent/blob/2983377abb1d332c438dbcf62b212ecdb0e5d68c/src/dict.php) I wrote somewhat recently.

**Comment by rtheunissen** [comment]
> Hash + linear search for collisions only

That's still O(n) though, I think. I like that idea though, it might not theoretically be equivalent but would be better in practice than it is currently.

**Comment by morrisonlevi** [comment]
It's linear for collisions only; it's constant time for everything else. How are you doing it in C?

**Comment by rtheunissen** [comment]
> It's linear for collisions only

Therefore a worse-case or "no worse than" linear, ie. O(n). Depending on the distributions of the hash values, in practice the performance would actually be pretty good though. :)

> How are you doing it in C?

Almost identically to the PHP hash table, except we don't case keys to string/int. We have a zval key and a zval value, and the collision chain is linked by the u2.next value of the key. Traversing the collision chain is linear in the same of the chain (so like.. O(1 + k)), but implementing this in PHP would require a full linear search from 0. It's also *very* heavy on memory because we still have to store the key/value pair as an array which has a minimum allocation of 8 zval's. It's not the end of the world, but you have to admit it's not ideal. 

See https://github.com/php-ds/extension/blob/master/src/ds/ds_htable.c#L336

**Comment by yybalam** [comment]
Hey. I came here to do a kind of apology to keep the polyfill. :)

You'll see, it is true that the polyfill implies a performance penalty, but also provide a nice *syntactic sugar* over PHP arrays.

I'd like to quote from the medium post: "**Performance should not be your top priority**. Code should be consistent, maintainable, robust, predictable, safe, and easy to understand. But that’s not to say that performance is '_not important_'."

That make me think, at most cases when you develop a *heavy load* application, you have access to the environment, the team at those cases are relatively *big*, so must every developer install the extension? I think this in the same way of the server. For development many team members do not install Nginx or Apache, with the PHP build in server it's enough. I love and use the pollyfill in that way, the performance penalty is negligible. Of course, for production a properly server and extension are installed.

By other hand, when a project do not suffer a *heavy load* we can be more flexible about performance topic, and yet get the benefit of the syntactic sugar. This provide a code that are more consistent and maintainable. And, if the project grows, the administrators can change for a scenario with a more accessible environment, so changing to that scenario, with installing the extension the project take an immediate advantage of performance.

Also, in my experience, the *big projects* have the more use cases for implement DS like Map or Set, even though the major DS to use is Vector. And with more modest project Vector become the number one DS. Maybe I'm wrong, but for me has been some like that.

Actually, could seem that for some project do an use of SplFixedArray instead of regular array is an option, but I think that Vector give more clean code, so I assume the performance penalty.

So, even I see there is some issues (like the #67), I think that for both scenarios the pollyfill can still be a useful tool, with the latent of an improvement.

Thank you for all the work. :)

**Comment by rtheunissen** [comment]
>  the team at those cases are relatively big, so must every developer install the extension?

Yes, in the same way that they would need to install other dependencies. It's as simple as having `pecl install ds` as part of your setup script or ideally part of your docker container.

> but also provide a nice syntactic sugar over PHP arrays.

While this is true, there are many other libraries written in PHP that offer the same thing. While performance isn't the top priority sometimes, it's an essential reason why this library exists and was written as an extension in the first place.

>  I love and use the pollyfill in that way, the performance penalty is negligible.

Why not just use the extension? You can add `ext-ds` to your composer file to indicate a requirement, and use pecl to install it. There's no cost and you don't get inconsistencies between developers who have it installed, those that do not, and your production environment. 

It's simply a risk that isn't worth facilitating.

**Comment by yybalam** [comment]
> Yes, in the same way that they would need to install other dependencies. It's as simple as having pecl install ds as part of your setup script or ideally part of your docker container.

It's true that is easy, but, for example, I code the same project sometimes at 3 different machines, so, in my Main machine I have installed the extension, but in  my secondary PC I'm fine with the polyfill. I know that is easy to install with PECL, but I'm a lazy person. :)

> While this is true, there are many other libraries written in PHP that offer the same thing. While performance isn't the top priority sometimes, it's an essential reason why this library exists and was written as an extension in the first place.

I will try to explain with an example: I develop an app for a customer who just had an Shared Hosting, so install the extension was impossible, but the app runs fine thank to the polyfill. Afetr, when the app base users grow, my customer hire a Cloud hosting, then I installed the extension and we get a performance improve immediately, no just by a more powerful server, also by the php-ds extension.

I was using another library instead of the pollyfill I would be to refactor for support php-ds. :)

> Why not just use the extension? You can add ext-ds to your composer file to indicate a requirement, and use pecl to install it. There's no cost and you don't get inconsistencies between developers who have it installed, those that do not, and your production environment.

Well, at the moment I did not have any inconsistency or trouble with my coworkers. O.O

Regards. :)

**Comment by rtheunissen** [comment]
Hi everyone. I've been exploring the idea of having only low-level structures as part of the extension, and implement the majority of the structures in PHP. The polyfill would still exist in this case.

The idea is that we can trade some raw performance for something that is more accessible, easier to maintain, and friendlier to contribute to. A big monolithic extension written in C can be intimidating and overwhelming for someone looking to contribute.

What do you all think?

**Comment by kelunik** [comment]
If your goal is a merge into the PHP core, everything in C is a must, I think. Otherwise the idea sounds good to me.

**Comment by rtheunissen** [comment]
> Maybe they would be open to only have some low-level structures in the core so that people could build their own libraries on top of them.

I think this right here is the primary goal. Once a library like this lands in core (or is used widely enough anyway), making updates to the extension will get harder. Updating a composer package that builds on it is a lot easier. The goal then becomes to make sure that the low level structures are solid and cover enough. 

How I have it set up now is to have the extension, a composer package for the polyfill that highly recommends installing the extension, and a composer package for the library which depends on the polyfill package. That way anyone can simply install php-ds/php-ds and optionally (definitely) install the extension to make use of the low level C implementations of the base structures.

I don't like knowing that we're potentially sacrificing some performance here but PHP is getting so efficient that the trade-off might not be as bad as it sounds. If we can manage the memory and actual data structures in C, building the API in PHP will be easy.

@morrisonlevi also very specific, small interfaces. :sparkles:

**Comment by rtheunissen** [comment]
My list of low level structures at this stage are:
- Deque (basic ring buffer, useful for anything operating on head or tail)
- Allocation (abstraction of a C array of zval's, can be reallocated, sliced, moved, etc)
- Binary Search Tree (Red/Black tree useful for things like sorted sets or searching)
- Binary Heap (optional comparator, min by default)
- HashTable (Stripped down API version of 1.0's Map, used for HashSet and HashMap)

I think that covers enough to get started. Am I missing anything fundamental?

I explored the idea of a ByteAllocation but I'm not sure that it would be consistent across platforms, or if anyone would ever want to allocate and work with data at the byte level.

**Comment by morrisonlevi** [comment]
I do not think it is in PHP's best interests nor this project's to move to core soon. As Rudi said:

> Once a library like this lands in core (or is used widely enough anyway), making updates to the extension will get harder.

I'm also an optimist about PHP's future and think we'll be getting nicer and nicer features over time. If we put this into core it is difficult to adopt such features e.g.

  - DS is not in core. To use the newer features and change API designs a DS v2.0 can be released (or v3.0, etc).
  - DS is in core. I highly doubt the PHP project would ship two concurrent major versions of DS in core, leading to a bad situation where we either upgrade it and break code, or we don't and we don't make progress.

----

As for implementing a portion in an extension and then building higher level concepts on top of that: I like the idea but am unsure what qualifies as sufficiently low-level and useful. Will have to thikn about that more.

**Comment by enumag** [comment]
> I think this right here is the primary goal. Once a library like this lands in core (or is used widely enough anyway), making updates to the extension will get harder. Updating a composer package that builds on it is a lot easier. The goal then becomes to make sure that the low level structures are solid and cover enough.

In that case I would recommend making 2.0 if this extension with the low-level structures and build a composer package on top of them as you said. Then have the community use it like that and focus on further improving and stabilizing the low-level structures before making a RFC to include them in PHP core.

I agree that moving to core soon is not necessarily a good idea, but it would be nice to prepare DS 2.0 with that in mind as eventual goal. Having to focus on and stabilize just a few low level structures should make this goal easier.

Yes, high level structures would lose some performance that way but in my opinion it's reasonable in this case. High level structures can be implemented in C later on as well.

**Comment by rtheunissen** [comment]
> I do not think it is in PHP's best interests nor this project's to move to core soon

I'm not too phased about this to be honest. If it stabilizes well and adoption is good, we can consider a proposal but at this point the benefit isn't obvious to be.

**Comment by designcise** [comment]
I'm not sure what the current state of development is regarding the entire discussion in here, but I landed here wanting to use the data structures in my microframework. My decision to use it is heavily based on the polyfill, because I would want the microframework to work for people who can't/don't want to install the extension. For people who want the performance gain, they would go the extra mile. If you're looking for greater adaptation, you have got to build for the bigger audience, and the bigger audience of PHP doesn't care for extensions all too much.

**Comment by rtheunissen** [comment]
We are actively planning ext-ds 2.0 on the 2.0 branch, and at this stage I do not foresee a polyfill implementation that mirrors the extension. I believe that this extension will only be successful if it is required, otherwise 3/5 projects will never actually install it and never truly depend on it or get any value from it. If you want to use them in your microframework, you should consider a hard dependency on the extension, or internal implementations of the interfaces.

What we could also do.. is provide a standard userland implementation of the interfaces, but have the classes *not* aliased. Eg. `Ds\Compat\HashMap implements Ds\Map`. It wouldn't be a "polyfill" technically but you could then use the interfaces in your microframework and allow the user to install the extension if they want.

I really, really want to avoid this though.

**Comment by designcise** [comment]
Totally understand. In that case, you should really stick to your priorities. Interfaces with a basic implementation sounds like a nice idea too, it might work to add a fallback when the extension doesn't exist.

When do you plan to release version 2? Do you have a date yet? And do you have anything new planned? Would love to hear about it.

**Comment by rtheunissen** [comment]
> When do you plan to release version 2? Do you have a date yet? And do you have anything new planned? Would love to hear about it.

Hopefully by the end of the year. If I could work on this full-time, maybe 3 months from now. I'm rewriting everything so don't expect much backward-compatibility. ^^

FWIW check out https://github.com/php-ds/ext-ds/blob/2.0/DRAFT.php

**Comment by kelunik** [comment]
@rtheunissen Does it make sense to have a separate extension name, so both versions could be installed at the same time?

**Comment by rtheunissen** [comment]
@kelunik I am open to that, but I haven't considered this. I think just a new major version is okay if we can somehow preserve the documentation from v1.

**Comment by kelunik** [comment]
@rtheunissen Allowing both versions being installed at the same time would allow for a smoother transition, because you can migrate things partially instead of all or nothing.

**Comment by rtheunissen** [comment]
Wouldn't that just cause more confusion? The namespace would have to change as well. I think either using one or the other is the best way forward. Following the migration guide should be easy enough to not require a temporary mixed state.

**Comment by kelunik** [comment]
Depends on how much changes, if you have libraries and own code depending on `ext-ds`, doing everything at once might not be that easy. But the namespace would have to change as well, yes.

**Comment by rtheunissen** [comment]
Supporting two versions of the same library at the same time is madness. :joy:

**Comment by glensc** [comment]
The existence of polyfill was the reason I even considered adopting `\Ds` in software I maintain. I can not control the environment my software is installed, and I do not want to limit installations (of my software) because there's a problem compiling the extension.

I'm ok if there's worse performance using the polyfill, if users (of my software) have the performance as an issue, they can install the extension. and I'm perfectly fine with that.

Thank you for maintaining both versions for now!

Perhaps the end goal should be to get support classes to php-core, like `ext-sodium`, until then maintain the code as external source.

(sorry I did not read the full thread here if this POV was already covered).

**Comment by hopeseekr** [comment]
> One thing to note is that the polyfill also acts as an autocomplete source. If we decide to abandon the polyfill in 2.0, we would need to keep the stubs anyway. Something we could do is maintain a composer package that contains the tests and the stubs, and _requires_ the extension as `ext-ds`, and have the extension repository require that package. That way we're still separating the extension from the tests and the stubs, and we're enforcing that the extension is installed if a project has the composer package as a dependency.

Look at my brand new PR at php-ds/polyfill#51: It's a PhpStorm stub. You could make that ENTIRE source code of record for the entire php-ds/php-ds composer package, completely abandon the php-ds/polyfill and STILL have 100% code completion + documentation.

In fact, once the PhpStorm stub is officially accepted by PhpStorm (Open a PR over at https://github.com/JetBrains/phpstorm-stubs), then that's EXACTLY what I recommend doing, professionally.

**Comment by tominventisbe** [comment]
I haven't seen this mentioned yet, but perhaps in PHP 7.4 it would become a possibility to use the FFI to implement this extension in PHP instead?

I realize the first version of the FFI does not have the performance of an actual extension, but IIRC there are plans to improve this (i.e. JIT in PHP 8).

**Comment by jkuchar** [comment]
FFI will be slower then native extension, but still with the same big-O complexity (just with added constant penalty for each function call, which is more then acceptable).

**Comment by glensc** [comment]
link for lazy ones about ffi that got merged: https://wiki.php.net/rfc/ffi

**Comment by rtheunissen** [comment]
@tominventisbe @jkuchar for a project like this (data structures being a special case), performance and low-level functionality is worth more than the benefit of using the FFI / implementing it in PHP. I don't know that for sure, but I would probably vote to keep the C implementation as the primary, and potentially implement it with FFI as well as an alternative option.

**Comment by designcise** [comment]
if you want wider adaptation then FFI looks like a really good option. It may be slow when it's introduced, but surely it will improve in newer versions

**Comment by morrisonlevi** [comment]
I don't understand how you are going to use FFI to "polyfill" this. Am I missing something? If you have the extension, then use it. If you don't have it, how are you going to call into it via FFI?

**Comment by rtheunissen** [comment]
> if you want wider adaptation then FFI looks like a really good option. It may be slow when it's introduced, but surely it will improve in newer versions

I disagree with this. No one else is calling for polyfill implementations of other extensions. If a library/extension provides enough value for enough people, it will be adopted by the community. Good examples are PDO, xdebug, gd/imagick, etc. We would be spending _a lot_ of time implementing and supporting an inferior version of something simply to avoid the need to install and enable an extension, which in my opinion indicates that the problem to solve is part of PHP's dependency management infrastructure, rather than extensions themselves.  

If an polyfill must exist, it won't concern itself with FFI.

**Comment by rtheunissen** [comment]
> I don't understand how to use FFI to "polyfill" this. If you have the extension, then use it. If you don't have it, how are you going to call into it via FFI? Am I missing something?

I read it as a rewrite/alternative version of the _main_ implementation, but instead of an extension it's written in PHP using native C bits and pieces? Anyone is welcome to attempt that, but it will not be an effort within the php-ds repo, at least from my side.

**Comment by glensc** [comment]
FFI is just a bridge to create bindings for shared libraries (`libfoo.so`) symbols, as there is no underlying library, FFI offers nothing here. for example imagick ext could be implemented as FFI, as there is exists `libMagickCore.so` and `libMagickWand.so`

**Comment by tominventisbe** [comment]
Yes, unfortunately, as I see it (though I may be missing something): if you also want to keep the extension around, you'd still be stuck maintaining two code bases as the FFI allows you to write extensions in PHP by binding to e.g. C libraries from PHP, as already mentioned by @glensc.

Since it's mostly real PHP code and some PHP wrapping/mapping the C functions, it's not native (e.g. C) code like a native extension would be, so you'd still need to maintain both.

You could use the FFI to load in a new php-ds C library instead (which would then also be used by the extension), but this then poses the problem of getting that new C library on your (shared) hosting. This could be done in some cases via static libraries - I believe there are some examples of this, such as `wkhtmltopdf` - but it's definitely not ideal.

I believe the most ideal course of action would be to "simply" (try and get) this extension merged in PHP itself as core extension. I think there is room for a good (better) set of container and data structures in PHP, though I'm not a core developer, who may think differently.

There may also be requirements, such as the extension having a certain amount of popularity if it is useful enough, such as mentioned by @rtheunissen. I must admit I'm afraid that this is not one of those types of extensions that is going to pull that amount of weight as it's not like databases where you may need to communicate with something a customer demands, and thus pretty much need the extension without question and get it installed.

This library is something that is usually seen as optional as there are some alternatives in pure PHP on Packagist (though arguably not as good and extensive, or we wouldn't be having this discussion :smile:), but there is little hard material to convince people to have such an extension installed on their servers - if that is at all possible and they are not on shared hosting - unless they require the performance, which is also a minority, I believe. If they don't require the performance, they are probably going to look for something not requiring any installation, especially since a custom extension also poses an additional security vector.

**Comment by hopeseekr** [comment]
If the polyfill is abandoned, so too goes the adoption. It's literally the only reason I'll risk distributing code w/ php-ds.

**Comment by rtheunissen** [comment]
> If the polyfill is abandoned, so too goes the adoption. It's literally the only reason I'll risk distributing code w/ php-ds.

This makes me sad, and I'm not sure if I agree, but I am open to the idea of a polyfill as long as it is *strongly* recommended that projects install the extension if it is at all possible to do so. 

I don't like the idea of potentially *slowing down* what could be most projects if the aversion to install an extension is shared by a majority. 

I'm just curious... if you are writing something that you would like to distribute (library, etc), and you'd like to use php-ds, why not require php-ds as a dependency? If a user can't install it, they also can't use your library. The problem to solve here is why they can't install it (or don't want to), and I worry that maintaining a polyfill encourages that aversion, and discourages the innovation we need around extension dependencies. 

Classic pragmatism vs. idealism situation here. Do we bite the bullet and maintain the polyfill for the sake of practicality, or do we insist that others should get with the times or be left behind?

Looking at 2.0, the API will almost definitely be trimmed *way* down, which would make the polyfill a lot easier to manage. If I had to vote on this today, I think I would vote to keep it because there is no clear consensus to abandon it. But if I ever hear of someone not installing the extension by choice... :policeman:

So until further notice, the decision here is to keep it. :tada:

**Comment by eduardbareev** [comment]
> convince people to have

Every php installation have some extensions installed be it `imagick`, `xdebug`, `uploadprogress` or whatever.
PHP historically requires installing native extensions. 

From what I see installing ds with `pecl install ds` doesn't require anything except usual build tools.

Popular ones are packaged by distro or third-party maintainer. I checked and I see that currently ds is in fedora repository and in remi repository too, so it's just  `dnf install php82-php-pecl-ds.x86_64`

Therefore for some installing an extension may be actually easier than installing an autoloaded composer polyfill.
